Rust 1.19 regression: cyclic reference detected for rusttype

Prior to `rustc 1.19.0-nightly (f1140a331 2017-05-08)`, [rusttype](https://crates.io/crates/rusttype) 0.2.1 could compile okay. Now it gives the following error:

```
   Compiling rusttype v0.2.1
error[E0391]: unsupported cyclic reference between types/traits detected
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:5
   |
47 |     keys: [K; CAPACITY],
   |     ^^^^^^^^^^^^^^^^^^^ cyclic reference
   |
note: the cycle begins when processing `support::bst::node::LeafNode::keys`...

# ... full output in the Details section below ...
```

*Steps to reproduce:*

```bash
cargo init mytest
cd mytest
echo 'rusttype = "0.2.1"' >> Cargo.toml
cargo build
```

A minimal example that reproduces this (playpen's nightly is still on `f4209651e 2017-05-05` so it compiles):

```rust
use std::ops::Mul;

const C: usize = 1;
const CAPACITY: usize = 1 * C;

struct A<X> {
    f: [X; CAPACITY],
}

struct B<T> {
    f: T,
}

impl<T> Mul for B<T> {
    type Output = Self;
    fn mul(self, _rhs: B<T>) -> Self::Output {
        self
    }
}

impl<T> Mul<usize> for B<T> {
    type Output = Self;
    fn mul(self, _rhs: usize) -> Self::Output {
        self
    }
}

fn main() {
    let a = A { f: [1] };
    let _ = B { f: a };
}
```

I have no idea why that reproduces it; I broke my brain coming up with that code.

My rustc version:

```
$ rustc --version --verbose
rustc 1.19.0-nightly (f1140a331 2017-05-08)
binary: rustc
commit-hash: f1140a33176a5fb2e91e26ea3ae42a834dd9bfdf
commit-date: 2017-05-08
host: x86_64-pc-windows-msvc
release: 1.19.0-nightly
LLVM version: 4.0
```

This is also seen on linux and OS X, such as in this [travis build](https://travis-ci.org/azriel91/autexousious/builds/230217063)

<details>

```
$ cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling odds v0.2.25
   Compiling linked-hash-map v0.0.10
   Compiling byteorder v0.4.2
   Compiling stb_truetype v0.2.1
   Compiling nodrop v0.1.9
   Compiling arrayvec v0.3.23
   Compiling rusttype v0.2.1
error[E0391]: unsupported cyclic reference between types/traits detected
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:5
   |
47 |     keys: [K; CAPACITY],
   |     ^^^^^^^^^^^^^^^^^^^ cyclic reference
   |
note: the cycle begins when processing `support::bst::node::LeafNode::keys`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:5
   |
47 |     keys: [K; CAPACITY],
   |     ^^^^^^^^^^^^^^^^^^^
note: ...which then requires const-evaluating `support::bst::node::LeafNode::{{initializer}}`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:15
   |
47 |     keys: [K; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires processing `support::bst::node::LeafNode::{{initializer}}`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:15
   |
47 |     keys: [K; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:15
   |
47 |     keys: [K; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which then requires processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which then requires coherence checking all impls of trait `std::ops::Mul`...
note: ...which then requires processing `geometry::Vector`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\geometry.rs:27:1
   |
27 | / pub struct Vector<N> {
28 | |     pub x: N,
29 | |     pub y: N
30 | | }
   | |_^
note: ...which then requires computing the variances for items in this crate...
   = note: ...which then again requires processing `support::bst::node::LeafNode::keys`, completing the cycle.

error[E0391]: unsupported cyclic reference between types/traits detected
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:15
   |
48 |     vals: [V; CAPACITY],
   |               ^^^^^^^^ cyclic reference
   |
note: the cycle begins when processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:47:15
   |
47 |     keys: [K; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which then requires processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which then requires coherence checking all impls of trait `std::ops::Mul`...
note: ...which then requires processing `geometry::Vector`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\geometry.rs:27:1
   |
27 | / pub struct Vector<N> {
28 | |     pub x: N,
29 | |     pub y: N
30 | | }
   | |_^
note: ...which then requires computing the variances for items in this crate...
note: ...which then requires processing `support::bst::node::LeafNode::vals`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:5
   |
48 |     vals: [V; CAPACITY],
   |     ^^^^^^^^^^^^^^^^^^^
note: ...which then requires const-evaluating `support::bst::node::LeafNode::{{initializer}}`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:15
   |
48 |     vals: [V; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires processing `support::bst::node::LeafNode::{{initializer}}`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:15
   |
48 |     vals: [V; CAPACITY],
   |               ^^^^^^^^
   = note: ...which then again requires processing `support::bst::node::CAPACITY`, completing the cycle.

error[E0391]: unsupported cyclic reference between types/traits detected
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cyclic reference
   |
note: the cycle begins when processing `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:44:1
   |
44 | pub const CAPACITY: usize = 2 * B - 1;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which then requires coherence checking all impls of trait `std::ops::Mul`...
note: ...which then requires processing `geometry::Vector`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\geometry.rs:27:1
   |
27 | / pub struct Vector<N> {
28 | |     pub x: N,
29 | |     pub y: N
30 | | }
   | |_^
note: ...which then requires computing the variances for items in this crate...
note: ...which then requires processing `support::bst::node::LeafNode::vals`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:5
   |
48 |     vals: [V; CAPACITY],
   |     ^^^^^^^^^^^^^^^^^^^
note: ...which then requires const-evaluating `support::bst::node::LeafNode::{{initializer}}`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:15
   |
48 |     vals: [V; CAPACITY],
   |               ^^^^^^^^
note: ...which then requires const-evaluating `support::bst::node::CAPACITY`...
  --> C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\rusttype-0.2.1\src\support\bst\node.rs:48:15
   |
48 |     vals: [V; CAPACITY],
   |               ^^^^^^^^
   = note: ...which then again requires processing `support::bst::node::CAPACITY`, completing the cycle.

error: aborting due to 3 previous errors

error: Could not compile `rusttype`.

To learn more, run the command again with --verbose.
```

</details>

--- a/src/librustc/infer/equate.rs
+++ b/src/librustc/infer/equate.rs
@@ -11,9 +11,12 @@
 use super::combine::{CombineFields, RelationDir};
 use super::{Subtype};
 
+use hir::def_id::DefId;
+
 use ty::{self, Ty, TyCtxt};
 use ty::TyVar;
-use ty::relate::{Relate, RelateResult, TypeRelation};
+use ty::subst::Substs;
+use ty::relate::{self, Relate, RelateResult, TypeRelation};
 
 /// Ensures `a` is made equal to `b`. Returns `a` on success.
 pub struct Equate<'combine, 'infcx: 'combine, 'gcx: 'infcx+'tcx, 'tcx: 'infcx> {
@@ -37,6 +40,22 @@ impl<'combine, 'infcx, 'gcx, 'tcx> TypeRelation<'infcx, 'gcx, 'tcx>
     fn tcx(&self) -> TyCtxt<'infcx, 'gcx, 'tcx> { self.fields.tcx() }
 
     fn a_is_expected(&self) -> bool { self.a_is_expected }
+
+    fn relate_item_substs(&mut self,
+                          _item_def_id: DefId,
+                          a_subst: &'tcx Substs<'tcx>,
+                          b_subst: &'tcx Substs<'tcx>)
+                          -> RelateResult<'tcx, &'tcx Substs<'tcx>>
+    {
+        // NB: Once we are equating types, we don't care about
+        // variance, so don't try to lookup the variance here. This
+        // also avoids some cycles (e.g. #41849) since looking up
+        // variance requires computing types which can require
+        // performing trait matching (which then performs equality
+        // unification).
+
+        relate::relate_substs(self, None, a_subst, b_subst)
+    }
 
     fn relate_with_variance<T: Relate<'tcx>>(&mut self,
                                              _: ty::Variance,

--- a/src/librustc/ty/relate.rs
+++ b/src/librustc/ty/relate.rs
@@ -51,6 +51,24 @@ pub trait TypeRelation<'a, 'gcx: 'a+'tcx, 'tcx: 'a> : Sized {
         Relate::relate(self, a, b)
     }
 
+    /// Relate the two substitutions for the given item. The default
+    /// is to look up the variance for the item and proceed
+    /// accordingly.
+    fn relate_item_substs(&mut self,
+                          item_def_id: DefId,
+                          a_subst: &'tcx Substs<'tcx>,
+                          b_subst: &'tcx Substs<'tcx>)
+                          -> RelateResult<'tcx, &'tcx Substs<'tcx>>
+    {
+        debug!("relate_item_substs(item_def_id={:?}, a_subst={:?}, b_subst={:?})",
+               item_def_id,
+               a_subst,
+               b_subst);
+
+        let opt_variances = self.tcx().variances_of(item_def_id);
+        relate_substs(self, Some(&opt_variances), a_subst, b_subst)
+    }
+
     /// Switch variance for the purpose of relating `a` and `b`.
     fn relate_with_variance<T: Relate<'tcx>>(&mut self,
                                              variance: ty::Variance,
@@ -107,25 +125,6 @@ impl<'tcx> Relate<'tcx> for ty::TypeAndMut<'tcx> {
             Ok(ty::TypeAndMut {ty: ty, mutbl: mutbl})
         }
     }
-}
-
-// substitutions are not themselves relatable without more context,
-// but they is an important subroutine for things that ARE relatable,
-// like traits etc.
-fn relate_item_substs<'a, 'gcx, 'tcx, R>(relation: &mut R,
-                                         item_def_id: DefId,
-                                         a_subst: &'tcx Substs<'tcx>,
-                                         b_subst: &'tcx Substs<'tcx>)
-                                         -> RelateResult<'tcx, &'tcx Substs<'tcx>>
-    where R: TypeRelation<'a, 'gcx, 'tcx>, 'gcx: 'a+'tcx, 'tcx: 'a
-{
-    debug!("substs: item_def_id={:?} a_subst={:?} b_subst={:?}",
-           item_def_id,
-           a_subst,
-           b_subst);
-
-    let opt_variances = relation.tcx().variances_of(item_def_id);
-    relate_substs(relation, Some(&opt_variances), a_subst, b_subst)
 }
 
 pub fn relate_substs<'a, 'gcx, 'tcx, R>(relation: &mut R,
@@ -291,7 +290,7 @@ impl<'tcx> Relate<'tcx> for ty::TraitRef<'tcx> {
         if a.def_id != b.def_id {
             Err(TypeError::Traits(expected_found(relation, &a.def_id, &b.def_id)))
         } else {
-            let substs = relate_item_substs(relation, a.def_id, a.substs, b.substs)?;
+            let substs = relation.relate_item_substs(a.def_id, a.substs, b.substs)?;
             Ok(ty::TraitRef { def_id: a.def_id, substs: substs })
         }
     }
@@ -308,7 +307,7 @@ impl<'tcx> Relate<'tcx> for ty::ExistentialTraitRef<'tcx> {
         if a.def_id != b.def_id {
             Err(TypeError::Traits(expected_found(relation, &a.def_id, &b.def_id)))
         } else {
-            let substs = relate_item_substs(relation, a.def_id, a.substs, b.substs)?;
+            let substs = relation.relate_item_substs(a.def_id, a.substs, b.substs)?;
             Ok(ty::ExistentialTraitRef { def_id: a.def_id, substs: substs })
         }
     }
@@ -372,7 +371,7 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
         (&ty::TyAdt(a_def, a_substs), &ty::TyAdt(b_def, b_substs))
             if a_def == b_def =>
         {
-            let substs = relate_item_substs(relation, a_def.did, a_substs, b_substs)?;
+            let substs = relation.relate_item_substs(a_def.did, a_substs, b_substs)?;
             Ok(tcx.mk_adt(a_def, substs))
         }
 

--- a/src/test/run-pass/issue-41849-variance-req.rs
+++ b/src/test/run-pass/issue-41849-variance-req.rs
@@ -1,0 +1,43 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Regression test for #41849.
+
+use std::ops::Mul;
+
+const C: usize = 1;
+const CAPACITY: usize = 1 * C;
+
+struct A<X> {
+    f: [X; CAPACITY],
+}
+
+struct B<T> {
+    f: T,
+}
+
+impl<T> Mul for B<T> {
+    type Output = Self;
+    fn mul(self, _rhs: B<T>) -> Self::Output {
+        self
+    }
+}
+
+impl<T> Mul<usize> for B<T> {
+    type Output = Self;
+    fn mul(self, _rhs: usize) -> Self::Output {
+        self
+    }
+}
+
+fn main() {
+    let a = A { f: [1] };
+    let _ = B { f: a };
+}

